### PR TITLE
feat(core,vendo): a Cloud key answers which companies a user is in

### PR DESCRIPTION
**Pull request by yousefh409**
Slice 1 of Tenant Organizations.

Ships inert — with no console route yet the GET 404s and the SDK reads it as no memberships, logged once. Host assertion still wins and never calls out.

## What lands

- **`@vendoai/core`** — `packages/core/src/tenant-directory.ts`, the one definition of the payload BOTH repos parse (`TenantDirectoryPayload`, `TenantLimits`, `TenantCap` + zod schemas). A cap carries its `scope`; no reader ever gets to assume one.
- **`cloudDirectory`** — the GET, cached 60s per subject. A 404/5xx/malformed 200 degrades to the stale entry or to no memberships, and never throws: this is awaited inside per-request context resolution, so a throw would be a 500 for the whole turn.
- **`tenantLimits`** — the console's caps on the limiter that already exists, off the SAME cached entry, so a cap costs zero extra requests. A user in no pool falls through and is allowed, because counting a pool you are not in throws and a throw is a deny with no message.
- **Composition** — `VENDO_API_KEY` fills the `memberships` seam only when the host left it wholly unset. An explicit `auth.memberships` wins and no client is ever constructed.
- **Server surface** — `cloudDirectory`, `tenantLimits`, `createLimiter` are exported from `@vendoai/vendo/server`; the console's cross-repo seam test drives the real consumer against them.

## Behavior change reviewers should see argued, not discover

**The boot summary now carries a sixth row on a keyed deployment.**

`boot-summary.ts:234` reads the *composed* `membershipsSeam`, and `:322` gates a `tenants` row on it. Before this branch a keyed deployment with no host assertion left that seam undefined, so the block carried five rows; it now carries six.

The row is correct, which is why the expectation moved rather than the code. `boot-summary.ts:318-321` states the rule it encodes: tenant connectors are per-ORG, and an org only ever reaches a request through the memberships seam, so without one no overlay can ever be selected. A Cloud-filled seam gives every keyed deployment exactly that, so per-org connectors really are serving at boot.

The alternative — gating the row on the host's own `config.auth?.memberships` — would have kept the block byte-identical at the cost of going silent about a live capability. That is a boot summary lying by omission, so it was rejected.

The test edit is isolated in its own commit (`fb14d784a`), touches one expectation, and deletes nothing.

## Verification

| Gate | Result |
|---|---|
| `pnpm build` | 0 |
| `pnpm typecheck` | 0 |
| `pnpm lint` | 0 (4 warnings, 0 errors, all pre-existing in `demo-bank`) |
| `packages/vendo` scoped suites | 5 files, 80 tests passed |
| `packages/core/tests/tenant-directory.test.ts` | 6 tests passed |

Each new behavior was proven by reverting it and watching exactly the intended test go red: the cap `scope` (1 red), the directory's fail-open `try`/`catch` (3 red), the `user.pools` guard (1 red), and the host-wins short-circuit (1 red).

The full suite is CI's job — the green `ci` check is the gate of record.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Resolves org memberships and tenant caps from Vendo Cloud when `VENDO_API_KEY` is set, and adds a top‑level `memberships` seam so `principal` hosts can opt out. Previously, keyed deployments without memberships showed five boot rows and caps used a rolling 1d/30d window; now Cloud fills memberships unless you assert them (boot summary adds a “tenants” row) and caps reset at UTC calendar boundaries.

- `@vendoai/core`: adds `TenantDirectoryPayload`, `TenantLimits`, `TenantCap`, and zod schemas.
- `@vendoai/vendo`: adds `cloudDirectory(...)` (GET `/api/v1/users/{externalId}/memberships`), cached 60s per subject; 404/5xx/malformed 200 degrade to stale or none and never throw.
- Composition: if the memberships seam is unset and `VENDO_API_KEY` is present, the seam is filled from `cloudDirectory`; an explicit `auth.memberships` or top‑level `memberships` wins outright and Cloud is never called.
- Limits: `tenantLimits(directory)` enforces console caps via the existing limiter off the same cached entry; `composeLimits` prefers `config.limits`, else the Cloud default. On a meterless store, Cloud caps do not compose and log a warning; an explicit policy on a meterless store still throws. Cap windows now reset at UTC midnight/day-1 and month start; a message at 23:59 no longer spends the next day’s allowance.
- Server surface: exports `cloudDirectory`, `tenantLimits`, and `createLimiter` from `@vendoai/vendo/server`. Docs list `memberships` in the handler options table.

- Review and rollout
  - Behavior change: keyed deployments now show a "tenants" row at boot because Cloud provides memberships; per‑org features are live.
  - To opt out, assert memberships: use top‑level `memberships` for `principal` hosts or `auth.memberships` for auth presets (e.g., `async () => []` to declare no orgs).
  - Ensure a usage meter to enforce console caps; otherwise caps are ignored with a warning.

<sup>Written for commit 316b2faa41880cd83e38f736720c6824186fe96c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/runvendo/vendo/pull/1526?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/vendo/src/tenant-limits.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Each cap's `scope` chooses the window and nothing else — `per-tenant` hands
+ * `count` the `org:<tenantId>` pool, `per-member` hands it nothing and counts
+ * the subject. Neither spelling is new grammar; the limiter already speaks both.
+ */
+
+/** Which cap governs which metered action, and the span each is counted over. */
+const CAPS: Record<LimitAction, { key: keyof TenantLimits; days: number; noun: string; span: string }> = {
+  message: { key: "messagesPerDay", days: 1, noun: "messages", span: "for today" },
+  generation: { key: "generationsPerMonth", days: 30, noun: "generations", span: "for this month" },
```
[the remark this answers is not in the document]
Comment (yousefh409) [reply packages/vendo/src/tenant-limits.ts]: Confirmed and fixed in 316b2faa4 — thank you, this was a real bug and your reproduction was right.

**Root cause.** The caps passed `{ days: 1 }` / `{ days: 30 }`, and `limits.ts:100-102` turns any `days`/`hours`/`minutes` into `new Date(Date.now() - lookback)` — a rolling window. The denial text promised "for today" / "for this month", so the label and the arithmetic genuinely disagreed: usage at 23:59 still consumed the next day's allowance at 00:01, and 30 days is not a calendar month.

**Fix.** `LimitWindow` already carried `since` (`core/src/limits.ts:26`) and the meter is since-native (`UsageCountQuery.since` is required, `core/src/store.ts:393`), so the windows now pass a boundary instead of an elapsed count:

- `messagesPerDay` → `{ since: start of the current UTC day }`
- `generationsPerMonth` → `{ since: start of the current UTC month }`

`pool` rides along unchanged, so `per-tenant` scope is unaffected.

**Why the labels stayed.** The wording was the correct half. A customer told "50 messages per day" expects a midnight reset, not a sliding 24 hours they cannot predict, and the spec's own language is messages/day + generations/month. So the fix makes the sentences true rather than retreating from them.

**On timezone.** UTC, deliberately, with no per-host knob — one deployment's users span zones, so there is no single local midnight to honour, and an invented default would reset quotas at an hour nobody chose. That constraint is stated in a comment at the boundary helpers.

**Tests.** Two boundary cases now drive the real limiter and the real reference meter — usage genuinely recorded at a past instant and genuinely filtered by `since`, with only the clock pinned so the boundary is a fixed point rather than whenever CI runs. Both were watched failing first against the `days:` windows, with exactly your symptom (`expected { allow: false } to deeply equal { allow: true }`), then green. A third case denies *inside* one period, so the first two cannot pass against a cap that had simply stopped counting.

File: packages/vendo/src/tenant-limits.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * resolved for this request, so a cap costs zero extra requests.
+ *
+ * Each cap's `scope` chooses the window and nothing else — `per-tenant` hands
+ * `count` the `org:<tenantId>` pool, `per-member` hands it nothing and counts
+ * the subject. Neither spelling is new grammar; the limiter already speaks both.
+ */
+
+/** Which cap governs which metered action, and the span each is counted over. */
+const CAPS: Record<LimitAction, { key: keyof TenantLimits; days: number; noun: string; span: string }> = {
+  message: { key: "messagesPerDay", days: 1, noun: "messages", span: "for today" },
```
[the remark this answers is not in the document]
Comment (yousefh409) [reply packages/vendo/src/tenant-limits.ts]: Confirmed and fixed in 316b2faa4 — thank you, this was a real bug and your reproduction was right.

**Root cause.** The caps passed `{ days: 1 }` / `{ days: 30 }`, and `limits.ts:100-102` turns any `days`/`hours`/`minutes` into `new Date(Date.now() - lookback)` — a rolling window. The denial text promised "for today" / "for this month", so the label and the arithmetic genuinely disagreed: usage at 23:59 still consumed the next day's allowance at 00:01, and 30 days is not a calendar month.

**Fix.** `LimitWindow` already carried `since` (`core/src/limits.ts:26`) and the meter is since-native (`UsageCountQuery.since` is required, `core/src/store.ts:393`), so the windows now pass a boundary instead of an elapsed count:

- `messagesPerDay` → `{ since: start of the current UTC day }`
- `generationsPerMonth` → `{ since: start of the current UTC month }`

`pool` rides along unchanged, so `per-tenant` scope is unaffected.

**Why the labels stayed.** The wording was the correct half. A customer told "50 messages per day" expects a midnight reset, not a sliding 24 hours they cannot predict, and the spec's own language is messages/day + generations/month. So the fix makes the sentences true rather than retreating from them.

**On timezone.** UTC, deliberately, with no per-host knob — one deployment's users span zones, so there is no single local midnight to honour, and an invented default would reset quotas at an hour nobody chose. That constraint is stated in a comment at the boundary helpers.

**Tests.** Two boundary cases now drive the real limiter and the real reference meter — usage genuinely recorded at a past instant and genuinely filtered by `since`, with only the clock pinned so the boundary is a fixed point rather than whenever CI runs. Both were watched failing first against the `days:` windows, with exactly your symptom (`expected { allow: false } to deeply equal { allow: true }`), then green. A third case denies *inside* one period, so the first two cannot pass against a cap that had simply stopped counting.

File: packages/vendo/src/tenant-limits.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (yousefh409) [reply packages/vendo/src/tenant-limits.ts]: Confirmed and fixed in 316b2faa4 — thank you, this was a real bug and your reproduction was right.

**Root cause.** The caps passed `{ days: 1 }` / `{ days: 30 }`, and `limits.ts:100-102` turns any `days`/`hours`/`minutes` into `new Date(Date.now() - lookback)` — a rolling window. The denial text promised "for today" / "for this month", so the label and the arithmetic genuinely disagreed: usage at 23:59 still consumed the next day's allowance at 00:01, and 30 days is not a calendar month.

**Fix.** `LimitWindow` already carried `since` (`core/src/limits.ts:26`) and the meter is since-native (`UsageCountQuery.since` is required, `core/src/store.ts:393`), so the windows now pass a boundary instead of an elapsed count:

- `messagesPerDay` → `{ since: start of the current UTC day }`
- `generationsPerMonth` → `{ since: start of the current UTC month }`

`pool` rides along unchanged, so `per-tenant` scope is unaffected.

**Why the labels stayed.** The wording was the correct half. A customer told "50 messages per day" expects a midnight reset, not a sliding 24 hours they cannot predict, and the spec's own language is messages/day + generations/month. So the fix makes the sentences true rather than retreating from them.

**On timezone.** UTC, deliberately, with no per-host knob — one deployment's users span zones, so there is no single local midnight to honour, and an invented default would reset quotas at an hour nobody chose. That constraint is stated in a comment at the boundary helpers.

**Tests.** Two boundary cases now drive the real limiter and the real reference meter — usage genuinely recorded at a past instant and genuinely filtered by `since`, with only the clock pinned so the boundary is a fixed point rather than whenever CI runs. Both were watched failing first against the `days:` windows, with exactly your symptom (`expected { allow: false } to deeply equal { allow: true }`), then green. A third case denies *inside* one period, so the first two cannot pass against a cap that had simply stopped counting.
